### ISSUE TITLE
Domain-based DFS Namespace

### ISSUE THREAD

**Issue by tak4625** [hierynomus/smbj#212]
Does it support Domain-based Namespace?

**Comment by hierynomus** [comment]
@tak4625 Have you tried it? In theory it should work as I've tried to follow the `MS-DFSC` spec. In practice I could've made a mistake :)

**Comment by tak4625** [comment]
I tried it, but have become the error. The test cord is follows.
```java
SmbConfig config = SmbConfig.builder()
    .withMultiProtocolNegotiate(true)
    .withSigningRequired(true)
    .withDfsEnabled(true)
    .build();
SMBClient client = new SMBClient(config);
Connection conn = client.connect("polaris.local");
Session session = conn.authenticate(new AuthenticationContext("nobunaga", "Corexx00".toCharArray(), "polaris.local"));
DiskShare share = (DiskShare) session.connectShare("PublicDocs");
System.out.println(share.list(""));
```

The following exception occurs. (The authentication seems to be success.)
```
Exception in thread "main" com.hierynomus.mssmb2.SMBApiException: STATUS_BAD_NETWORK_NAME(3221225676/3221225676): Could not connect to \\polaris.local\PublicDocs
	at com.hierynomus.smbj.session.Session.connectTree(Session.java:113)
	at com.hierynomus.smbj.session.Session.connectShare(Session.java:98)
	at jp.co.hogehoge.gw.fm.DfsTest.main(DfsTest.java:25)
```

AD domain is "polaris.local" and domain-based DFS namespace is "\\\\polaris.local\PublicDocs".
I can connect to DFS by Windows Explorer.

**Comment by git-aleksey** [comment]
I'm having the same issue. What I found is that instead of connecting to the DFS share, it resolves the domain name to the domain controller and tries to find the share there.
So:
\\vn.com\Share is resolved to \\dc\Share  where dc is the domain controller for the domain. Connection fails with STATUS_BAD_NETWORK_NAME as it can't find the share.

**Comment by pepijnve** [comment]
@git-aleksey I tried reproducing this issue today but it seems to work fine for me. Could you describe the structure of your DFS namespace in a bit more detail?

> it resolves the domain name to the domain controller and tries to find the share there.

That's actually how the DFS client is supposed to work as far as I can tell, but I might be overlooking some subtlety in the spec.

**Comment by git-aleksey** [comment]
@pepijnve I setup a simple testing environment in Azure:  virtual network consisting of 3 servers (all Server 2012 R2) one of which acts as the domain controller.  I configured a domain-based dfs namespace with the root target on one of the servers:
vn.com – domain name
dc – Domain controller for vn.com
Share – DFS root
server1 – root target server 
So the dfs root for the namespace is at \\vn.com\Share with the link target at \\server1\Share

As far as I understand the process, vn.com should refer to the domain controller (dc) and retrieve the dfs metadata for the namespace from AD. The metadata should reference the root target (\\server1\Share). I’m not sure where the client kicks in here though.
From what I can see it gets to the domain controller but instead of querying AD it just processes it as a stand-alone namespace i.e. trying to access the share at \\dc\Share

Apologies if I misuse the terminology, DFS is not in my area of competence.

Thanks

**Comment by pepijnve** [comment]
Ok, that explains why it worked for me. I had configured a namespace on a DC and was using the DC as root server as well.

**Comment by git-aleksey** [comment]
Great. Is this something you are thinking to add to the next release or there is a backlog?

**Comment by hierynomus** [comment]
@git-aleksey It would really help to have access to a testing environment for this :).

We would need to see which step of the MS-DFSC spec it wrongly interpreted/implemented. Can we use your testing environment?

**Comment by pepijnve** [comment]
@hierynomus do you or your company have access to an msdn subscription? If you can spin up a couple of windows server VMs it'll be pretty easy to reproduce.

**Comment by git-aleksey** [comment]
Sure. just looking at it now. Will set it all up and drop you the details on jeroen's email.

**Comment by pepijnve** [comment]
@git-aleksey I was able to reproduce what you're seeing locally and have a first draft of a fix ready in the referenced PR. In short, smbj was missing a small part of the DFS puzzle when the server hosting the DFS namespace is not the DC.

**Comment by git-aleksey** [comment]
@pepijnve Great! When do you think we will have a release with the feature in? Most of our customers itching to turn SMB1 off. Would like to give them some timescales.
Thanks guys btw. Awesome job with the library!

**Comment by pepijnve** [comment]
> When do you think we will have a release with the feature in?

 ¯\_(ツ)_/¯ That's up to @hierynomus

**Comment by hierynomus** [comment]
@tak4625 I've merged #256. This should fix this issue, so I'm closing it. Feel free to re-open or open a followup ticket is something is still amiss

**Comment by tak4625** [comment]
@hierynomus Thank you. I'll give it a try.
